app/abci.go

```diff
@@ -362,6 +389,70 @@ func (app *HeimdallApp) PreBlocker(ctx sdk.Context, req *abci.RequestFinalizeBlo
 		return nil, errors.New("no validators found")
 	}
 
+	majorityMilestone, aggregatedProposers, proposer, err := milestoneAbci.GetMajorityMilestoneProposition(ctx, validators, extVoteInfo, logger)
+	if err != nil {
+		logger.Error("Error occurred while getting majority milestone proposition", "error", err)
+		return nil, err
+	}
+
+	if majorityMilestone != nil {
+		params, err := app.ChainManagerKeeper.GetParams(ctx)
+		if err != nil {
+			logger.Error("Error occurred while getting chain manager params", "error", err)
+			return nil, err
+		}
+
+		hasMilestone, err := app.MilestoneKeeper.HasMilestone(ctx)
+		if err != nil {
+			logger.Error("Error occurred while checking for the last milestone", "error", err)
+			return nil, err
+		}
+
+		addMilestone := func() error {
+			addMilestoneCtx, msCache := app.cacheTxContext(ctx)
+
+			logger.Debug("Adding milestone", "hash", common.Bytes2Hex(majorityMilestone.BlockHash), "number", majorityMilestone.BlockNumber)
+
+			if err := app.MilestoneKeeper.AddMilestone(addMilestoneCtx, milestoneTypes.Milestone{
+				Proposer:    proposer,
+				Hash:        majorityMilestone.BlockHash,
+				StartBlock:  majorityMilestone.BlockNumber,
+				EndBlock:    majorityMilestone.BlockNumber,
+				BorChainId:  params.ChainParams.BorChainId,
+				MilestoneId: common.Bytes2Hex(aggregatedProposers),
+				Timestamp:   uint64(ctx.BlockHeader().Time.Unix()),
+			}); err != nil {
+				logger.Error("Error occurred while adding milestone", "error", err)
+				return err
+			}
+
+			msCache.Write()
+			return nil
+		}
+
+		if hasMilestone {
+			lastMilestone, err := app.MilestoneKeeper.GetLastMilestone(ctx)
+			if err != nil {
+				logger.Error("Error occurred while fetching the last milestone", "error", err)
+				return nil, err
+			}
+
+			if lastMilestone.EndBlock+1 == majorityMilestone.BlockNumber {
```

Comment:
Initially I added `BlockNumber` to milestone proposition but then I removed it because I realized if we implement our logic right, it should not be possible to try to finalize milestone out of order.

Reply:
I added it mostly for debugging purpose. Doesn't hurt to keep this check unless it has significant impact on the performance.

Reply:
My concern is to not have bug that we could ignore because of this check. If we keep this can we atleast change the logger.Info in the else to logger.Error with big caps, we do that on other places for edge cases that should not happen.

Reply:
I will change the log level to warning. Ideally this shouldn't happen.

Reply:
Actually I think it can happen, if lets say at block N+1, 6 blocks have passed and we haven't finalized any new milestone, we will propose new milestone starting from the last finalized one but from block N we have proposed other milestone which is more forward, so if it gets 2/3 of votes in the PreBlocker of N+1 we gonna try to finalize it out of order. So we really need the check I think.

Reply:
If I understood this particular scenario correctly, at N+2 we'll anyway find  common ancestors with the milestone finalized at N+1 right ? So basically we ignore them and finalize from last milestone block at N+1.

Reply:
If we propose 1, 2, 3 at block N, then we propose 4, 5, 6 at N+1 but we finalize only 1, 2 at N+1. At N+2  we wont be able to finalize 4, 5, 6.